core/trino-main/src/main/java/io/trino/server/NodeStateManager.java

```diff
@@ -69,61 +75,114 @@ public NodeStateManager(
 
-    public synchronized void requestShutdown()
+    public NodeState getServerState()
     {
-        log.info("Shutdown requested");
+        return nodeState.get();
+    }
 
-        if (isCoordinator) {
-            throw new UnsupportedOperationException("Cannot shutdown coordinator");
-        }
+    /*
+    Below is a diagram with possible states and transitions
+
+    @startuml
+    [*] --> ACTIVE
+    note "state INACTIVE is not used internally\nis only used when the service is unavailable " as a
+    ACTIVE --> SHUTTING_DOWN : shutdown
+    ACTIVE --> DRAINING : drain
+    DRAINING --> ACTIVE: reactivate
+    DRAINING --> DRAINED
+    DRAINING --> SHUTTING_DOWN : gracefulShutdown
+    DRAINED --> ACTIVE: reactivate
+    DRAINED --> SHUTTING_DOWN : terminate
+    SHUTTING_DOWN --> [*]
+    @enduml
 
-        if (shutdownRequested) {
+    NOTE: SHUTTING_DOWN is treated as one-way transition to be 100% backwards compatible.
+    */
+    public synchronized void transitionState(NodeState state)
+    {
+        NodeState currState = nodeState.get();
+        if (currState == state) {
             return;
         }
-        shutdownRequested = true;
 
-        // wait for a grace period (so that shutting down state is observed by the coordinator) to start the shutdown sequence
-        shutdownHandler.schedule(this::shutdown, gracePeriod.toMillis(), MILLISECONDS);
+        switch (state) {
+            case ACTIVE -> {
+                if (currState == DRAINING && nodeState.compareAndSet(DRAINING, ACTIVE)) {
+                    return;
+                }
+                if (currState == DRAINED && nodeState.compareAndSet(DRAINED, ACTIVE)) {
+                    return;
+                }
+            }
+            case SHUTTING_DOWN -> {
+                if (currState == DRAINED && nodeState.compareAndSet(DRAINED, SHUTTING_DOWN)) {
+                    requestTerminate();
+                    return;
+                }
+                requestGracefulShutdown();
+                nodeState.set(SHUTTING_DOWN);
+                return;
+            }
+            case DRAINING -> {
+                if (currState == ACTIVE && nodeState.compareAndSet(ACTIVE, DRAINING)) {
+                    requestDrain();
+                    return;
+                }
+            }
+            case DRAINED -> throw new IllegalStateException(format("Invalid state transition from %s to %s, transition to DRAINED is internal only", currState, state));
+
+            case INACTIVE -> throw new IllegalStateException(format("Invalid state transition from %s to %s, INACTIVE is not a valid internal state", currState, state));
+        }
+
+        throw new IllegalStateException(format("Invalid state transition from %s to %s", currState, state));
     }
 
-    private void shutdown()
+    public synchronized void requestDrain()
     {
-        List<TaskInfo> activeTasks = getActiveTasks();
+        log.info("Drain requested, NodeState: " + getServerState());
```

Comment:
log.debug

Reply:
you are right, the io.trino.server.ServerInfoResource#updateState already logs any requests (should be a rare event), so anything here should be level debug or removed. 
Also, I think I can lower the method visibility to private